// zod/lib/src/types/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var parser_1 = require("../parser");
// import { maskUtil } from '../helpers/maskUtil';
// import { Masker } from '../masker';
// import { ZodString } from './string';
// import { maskUtil } from '../helpers/maskUtil';
var ZodTypes;
(function (ZodTypes) {
    ZodTypes["string"] = "string";
    ZodTypes["number"] = "number";
    ZodTypes["boolean"] = "boolean";
    ZodTypes["date"] = "date";
    ZodTypes["undefined"] = "undefined";
    ZodTypes["null"] = "null";
    ZodTypes["array"] = "array";
    ZodTypes["object"] = "object";
    // interface = 'interface',
    ZodTypes["union"] = "union";
    ZodTypes["intersection"] = "intersection";
    ZodTypes["tuple"] = "tuple";
    ZodTypes["record"] = "record";
    ZodTypes["function"] = "function";
    ZodTypes["lazy"] = "lazy";
    ZodTypes["lazyobject"] = "lazyobject";
    ZodTypes["literal"] = "literal";
    ZodTypes["enum"] = "enum";
})(ZodTypes = exports.ZodTypes || (exports.ZodTypes = {}));
//   interface Assertable<T> {
//     is(value: any): value is T;
//     assert(value: any): asserts value is T;
// }
var ZodType = /** @class */ (function () {
    // mask = <P extends maskUtil.Params<Type>>(_params: P): ZodType<maskUtil.Pick<Type, P>> => {
    //   return Masker(this, _params) as any;
    // };
    // pick = <Params extends maskUtil.Params<Type>>(_params: Params): maskUtil.Mask<Type, Params> => {
    //   return 'asdf' as any;
    // };
    function ZodType(def) {
        this.parse = parser_1.ZodParser(def);
        this._def = def;
    }
    ZodType.prototype.is = function (u) {
        try {
            this.parse(u);
            return true;
        }
        catch (err) {
            return false;
        }
    };
    ZodType.prototype.check = function (u) {
        try {
            this.parse(u);
            return true;
        }
        catch (err) {
            return false;
        }
    };
    return ZodType;
}());
exports.ZodType = ZodType;
